Support CREATE FUNCTION execution

caithagoras opened the pull request:
```
== RELEASE NOTES ==

General Changes
* ...
* ...
```

caithagoras commented:
Ready for initial review. Will add more tests for commit `Support MySQL based function namespace manager`

rongrong commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java:
```diff
@@ -169,23 +171,29 @@ public void registerBuiltInFunctions(List<? extends BuiltInFunction> functions)
         }
         else {
-            functionName = QualifiedFunctionName.of(name.getOriginalParts());
+            checkArgument(name.getOriginalParts().size() == 3, "Require 3 parts for qualified function name");
```
This should throw a `PrestoException` with proper error message that's easy to understand for user. Something along the line of `Reference to non-builtin functions must be using qualified function name in the form of "a.c.b"` if `originalParts = 2`, and throwing `FUNCTION_NOT_FOUND` is reasonable if `originalParts > 3`. The 3 parts function name should be a side effect of how function namespaces are managed. Since all sql function names would have 3 parts, a >3 parts function can not be resolved.

rongrong commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java:
```diff
@@ -219,13 +227,15 @@ private FunctionHandle resolveFunction(Optional<Session> session, QualifiedName
     {
-        return functionNamespaces.get(functionHandle.getFunctionNamespace()).getFunctionMetadata(functionHandle);
+        Optional<FunctionNamespaceManager<?>> functionNamespaceManager = getServingFunctionNamespaceManager(functionHandle.getFunctionNamespace());
+        checkArgument(functionNamespaceManager.isPresent(), "Cannot find function namespace for '%s'", functionHandle.getFunctionNamespace());
```
We already have a function handle here, how could there be no FunctionNamespaceManager? Why do we need this new logic?

caithagoras replied:
The `functionNamespaces` variables stores the mapping from prefix to manager, and it is acceptable that the prefix only contains the catalog.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/SqlInvokedFunction.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface SqlInvokedFunction
```
What are the other potential kind of `SqlInvokedFunction` other than `SqlInvokedRegularFunction`?

caithagoras replied:
Convert `SqlInvokedFunction`, `SqlFunctionId`, and `SqlFunctionHandle` to concrete classes.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionId.java:
```diff
@@ -0,0 +1,21 @@
+{
+    QualifiedFunctionName getFunctionName();
+
+    String getValue();
```
This is vague, what does "value" mean here?

caithagoras replied:
Renamed to `getId()`. This method effectively returns `toString`, the String representation of the `FunctionId`.

rongrong commented on presto-main/src/test/java/com/facebook/presto/metadata/TestFunctionManager.java:
```diff
@@ -79,7 +65,7 @@ public void testIdentityCast()
     {
         TypeRegistry typeManager = new TypeRegistry();
-        FunctionManager functionManager = createFunctionManager(typeManager);
+        FunctionManager functionManager = new FunctionManager(typeManager, new BlockEncodingManager(typeManager), new FeaturesConfig());
```
What is this change about?

caithagoras replied:
This belongs to a different PR.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,176 @@
+            "  `language` varchar(64) NOT NULL,\n" +
+            "  `deterministic` tinyint(1) NOT NULL,\n" +
+            "  `called_on_null_input` tinyint(1) NOT NULL,\n" +
+            "  `body` mediumtext,\n" +
```
I think text is enough. Not sure in what situation we would want function implementations to be > 64KB.

caithagoras replied:
Removed for now.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,176 @@
+            "  `created_time` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,\n" +
+            "  `modified_time` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,\n" +
+            "  PRIMARY KEY (`id`),\n" +
+            "  UNIQUE KEY `function_id_version` (`function_id`,`version`),\n" +
```
Semantically `function_id` should be a function name + parameters types. However this is not enforced on the tables. So if function_id does not have the same uniqueness as function name + parameters this implementation is in trouble.

caithagoras replied:
Removed for now.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,176 @@
+            "  `id` bigint(20) NOT NULL AUTO_INCREMENT,\n" +
+            "  `function_id` varchar(700) NOT NULL,\n" +
+            "  `version` bigint(20) unsigned NOT NULL,\n" +
+            "  `function_name` varchar(512) NOT NULL,\n" +
```
a 512 character function name sounds way too long. Seems like this would include the function namespace as well?

caithagoras replied:
Yes, it includes the function namespace.

Commit removed for now.

rongrong commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,176 @@
+            ") v\n" +
+            "    ON t.`function_id` = v.`function_id`\n " +
+            "    AND t.`version` = v.`version`")
+    List<SqlInvokedFunction> listFunctions(@Define("table_name") String tableName);
```
I think `listFunctions` only need to return a `SqlFunction`. This is used for `SHOW FUNCTIONS`, which only needs the metadata, not the body, or the version. If you cannot find a good class, maybe just return `List<FunctionMetadata>`.

caithagoras replied:
This `listFunctions` is at the end of the call stack for `Metadata.listFunction()`. We'll need refactor
a few layers of methods in order to do that, including `Metadata.listFunction()`, `FunctionManager.listFunction()`, `FunctionNamespaceManager.listFunction()`. For now, I'm adding a TODO in `FunctionNamespaceManager.listFunction()`

caithagoras commented:
@rongrong Thanks for the review! `Support MySQL based function namespace manager` is the last commit. The commits beyond have not been finished and will be moved to a separate PR for the purpose of executing CREATE FUNCTION.

caithagoras commented on presto-spi/src/main/java/com/facebook/presto/spi/function/CatalogSchemaPrefix.java:
```diff
@@ -0,0 +1,90 @@
+    private final String catalogName;
+    private final Optional<String> schemaName;
+
+    public CatalogSchemaPrefix(String catalogName, Optional<String> schemaName)
```
This class is `CatalogSchemaPrefix`. It is used to represent `catalog` or `catalog.schema`. The current usage in this commit is represent the list of "serving namespaces" of a function namespace manager. Example:
```
serving-namespaces=testing,dev.di
```
A "serving namespaces" is basically `List<CatalogSchemaPrefix>`

caithagoras commented on presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java:
```diff
@@ -13,168 +13,75 @@
 
-    public String getSuffix()
+    // TODO: Examine all callers to limit the usage of the method
+    public String getUnqualifiedName()
```
I do think `suffix` is accurate in certain cases. For example, it makes sense to say `c` is the suffix of a String `a.b.c`. It also make sense to say `c` is the suffix of a `QualifiedName("a.b.c")`  where it is consisted of `List<String> parts`.

However, we're explicitly defining a `QualifiedFunctionName` in terms of `(CatalogSchemaName functionNamespace, String unqualifiedName)`. Also, using this name indicates an issue. Eventually, we'll be able to remove this method completedly.

rongrong replied:
That should be defined as `(CatalogSchemaName functionNamespace, String suffix)` then.

caithagoras commented on presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java:
```diff
@@ -0,0 +1,176 @@
+{
+    @SqlUpdate("CREATE TABLE `<table_name>` (\n" +
+            "  `id` bigint(20) NOT NULL AUTO_INCREMENT,\n" +
+            "  `function_id` varchar(700) NOT NULL,\n" +
```
This is because the max key length for MySQL is 768.

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionNamespaceManager.java:
```diff
@@ -30,15 +30,15 @@
      * {@link #commit(FunctionNamespaceTransactionHandle)} is called.
      */
-    void rollback(FunctionNamespaceTransactionHandle transactionHandle);
+    void abort(FunctionNamespaceTransactionHandle transactionHandle);
```
Should this be a different commit?

caithagoras replied:
Done

rongrong commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java:
```diff
@@ -331,20 +343,20 @@ private FunctionHandle lookupFunction(
-                servingFunctionNamespaceManager = functionNamespace.getValue();
+        for (Map.Entry<CatalogSchemaPrefix, FunctionNamespaceManager<?>> entry : functionNamespaces.entrySet()) {
+            CatalogSchemaPrefix prefix = entry.getKey();
+            if (prefix.includes(functionNamespace) && (bestMatch == null || bestMatch.includes(prefix))) {
```
Maybe have a `includesButNotEqual` instead. Currently this is relying on same prefix would not appear more than once.

caithagoras replied:
I'm not sure what will be the behavior of the `includesButNotEqual` method you mentioned here.

rongrong replied:
the `includes` method in prefix returns true of it has a bigger or equal scope. I'm suggesting a method that would just return true if it actually has a bigger scope. For example `"a.b" includes "a.b" => true`, `"a.b" includes but not equal "a.b" => false`.

rongrong commented on presto-main/src/main/java/com/facebook/presto/sql/gen/FunctionCallCodeGenerator.java:
```diff
@@ -49,7 +49,7 @@ public BytecodeNode generateCall(FunctionHandle functionHandle, BytecodeGenerato
 
         return context.generateCall(
-                functionManager.getFunctionMetadata(functionHandle).getName().getSuffix(),
+                functionManager.getFunctionMetadata(functionHandle).getName().getUnqualifiedName(),
```
I think I mentioned before that `getSuffix` is more accurate than `getUnqualifiedName`. If you prefer to use "unqualified name", it's more appropriate to have a utility function `unqualifyName(qualifiedName)`. I'm not sure whether you can call the last part of a qualified name "unqualified name".

caithagoras replied:
It is a convention to called the last part of a "fully qualified object" (file path, class name, etc) a unqualified name. Here are some reference:
1. https://stackoverflow.com/a/27966682
2. https://en.wikipedia.org/wiki/Fully_qualified_name

rongrong replied:
Again, those reference only said a name without prefix is a unqualified name, rather than saying the last part of a qualified name is a "unqualified name". The reference you mentioned also used "suffix", which I still think is better than "unqualified name". If you want to be stubborn on this I won't insist. But I think "suffix" is better than "unqualified name" in this case.

caithagoras replied:
Looking at some of our existing classes:
- `QualifiedObjectName` is defined as (catalogName, schemaName, objectName)
- `QualifiedTablePrefix` is defined as (catalogName, schemaName, tableName)

What about defining `QualifiedFunctionName` is (functionNamespace, functionName)?

rongrong commented on presto-spi/src/main/java/com/facebook/presto/spi/function/SqlFunctionHandle.java:
```diff
@@ -11,49 +11,35 @@
-import static java.util.stream.Collectors.joining;
 
-public class SqlInvokedRegularFunctionHandle
+public class SqlFunctionHandle
```
Add `@Experimental` to all classes introduced to spi.

rongrong commented on presto-sql-function/src/main/java/com/facebook/presto/sqlfunction/AbstractSqlInvokedFunctionNamespaceManager.java:
```diff
@@ -127,6 +127,11 @@ public final FunctionMetadata getFunctionMetadata(FunctionHandle functionHandle)
         return metadataByHandle.getUnchecked((SqlFunctionHandle) functionHandle);
     }
 
+    protected void refreshFunctionsCache(QualifiedFunctionName functionName)
```
Where is this used?

caithagoras replied:
It is used by subclasses in which `createFunction` is implemented. This method can be called to update the functions cache on a new function or a new version fo a function is created.

caithagoras replied:
Yes, it is in `presto-facebook` now that we have remove the plugin commit from the PR.

rongrong commented on presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java:
```diff
@@ -0,0 +1,104 @@
+        return immediateFuture(null);
+    }
+
+    private SqlInvokedFunction createFunctionToSqlInvokedFunction(CreateFunction statement)
```
the function name is a little strange. Maybe `createSqlInvokedFunction`?

rongrong replied:
This comment was not addressed.

rongrong commented on presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java:
```diff
@@ -0,0 +1,104 @@
+    private SqlInvokedFunction createFunctionToSqlInvokedFunction(CreateFunction statement)
+    {
+        if (statement.getFunctionName().getParts().size() != 3) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Invalid SqlFunction name: %s, require exactly 3 parts", statement.getFunctionName().getParts().size()));
```
Should this be an user error?

rongrong commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java:
```diff
@@ -343,6 +347,15 @@ private FunctionHandle lookupFunction(
+    {
+        Optional<FunctionNamespaceManager<?>> functionNamespaceManager = getServingFunctionNamespaceManager(functionNamespace);
+        if (!functionNamespaceManager.isPresent()) {
+            throw new PrestoException(FUNCTION_NOT_FOUND, format("Cannot find function namespace '%s'", functionNamespace));
```
If this is called during `CREATE FUNCTION` then this error code is not appropriate. Maybe this class should just throw run time error and let caller decide what error code it should use.

caithagoras replied:
Since different callers need to handle them differently, I'm removing this method and handle the cases of `!functionNamespaceManager.isPresent()` in each caller (2) separately.

rongrong commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -538,6 +540,26 @@ protected Scope visitCreateView(CreateView node, Optional<Scope> scope)
+                    .withRelationType(RelationId.anonymous(), new RelationType(fields))
+                    .build();
+            Type bodyType = analyzeExpression(node.getBody(), functionScope).getExpressionTypes().get(NodeRef.of(node.getBody()));
+            metadata.getTypeManager().canCoerce(bodyType, returnType);
```
I'm not sure whether there's implications at run time if the return type of body is not exactly the same as the given `returnType`. Is there valid reasons that we want to allow coercion? Why not just require the return type to be exact?

caithagoras replied:
This is to add more flexibility to the SQL functions Presto support. I can change it to be the exact match, and properly test for the coerced case before rolling out.

rongrong replied:
Let's do exact match for now. It's easier to relax constraints than add them in the future.

rongrong replied:
You didn't use the return value btw.

caithagoras replied:
Done

rongrong commented on presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java:
```diff
@@ -0,0 +1,104 @@
+                RoutineCharacteristics.Language.valueOf(statement.getCharacteristics().getLanguage().name()),
+                RoutineCharacteristics.Determinism.valueOf(statement.getCharacteristics().getDeterminism().name()),
+                RoutineCharacteristics.NullCallClause.valueOf(statement.getCharacteristics().getNullCallClause().name()));
+        String body = formatSql(statement.getBody(), Optional.empty());
```
It's a little strange to see `formatSql` here. Makes more sense to call it in `SHOW CREATE FUNCTION` or something like that. Consider that we didn't really do a good job in formatting expression, the original user input might even be better in a lot of cases.

caithagoras replied:
That's the only non-hacky way to convert an AST to a String representation as far as I know.

caithagoras commented:
Added explainDdl test.

caithagoras commented:
- Added tests in `TestSqlFunctions`.
- Added semantic checks for duplicating column names.
- Added TODO to ensure no SQL invoked functions in function body.
- Fix an issue for fetching function implementation in FunctionManager.

rongrong commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java:
```diff
@@ -49,6 +49,7 @@
     INVALID_LITERAL,
 
     FUNCTION_NOT_FOUND,
+    DUPLICATE_PARAMETER_DECLARATION,
```
To be consistent with `DUPLICATE_COLUMN_NAME`, maybe name this to `DUPLICATE_PARAMETER_NAME`?

caithagoras commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -538,6 +547,45 @@ protected Scope visitCreateView(CreateView node, Optional<Scope> scope)
+            analysis.setUpdateType("CREATE FUNCTION");
+
+            // Check parameter
+            List<String> duplicateParameters = node.getParameters().stream()
```
I need the duplicate parameter names to construct a user-friendly error message, but size checking won't tell me what are duplicated.

Alternatively, printing one name instead of a list is fine with me as well, but that does not make the code simpler.

rongrong replied:
I'd just print the whole parameter list lol but sure.

rongrong commented:
Thanks for working on this! Looking forward to see this in production!
